src/irr/asset/CCPUMeshPacker.h

```diff
@@ -46,85 +67,243 @@ PackedMeshBufferData CCPUMeshPacker<MDIStructType>::alloc(const ICPUMeshBuffer c
 
-	for (int i = 0; i < MDIStructsNeeded; i++)
+	const uint64_t currMeshBufferIdxCnt = inputMeshBuffer->getIndexCount();
+	const size_t MDIStructsNeeded = core::roundUp<size_t>(currMeshBufferIdxCnt, m_maxTriangleCountPerMDIData);
+
+	uint32_t firstIdxForCurrMeshBatch = ramb.indexAllocationOffset;
+
+	//assuming it's never EIT_UNKNOWN
+	IndexType* idxBuffBatchBegin = static_cast<IndexType*>(inputMeshBuffer->getIndexBufferBinding()->buffer->getPointer());
+	IndexType* idxBuffBatchEnd = nullptr;
+
+	/*struct ReservedAllocationMeshBuffers
 	{
-		uint32_t idxCnt = 0u;
+		uint32_t mdiAllocationOffset;
+		uint32_t mdiAllocationReservedSize;
+		uint32_t instanceAllocationOffset;
+		uint32_t instanceAllocationReservedSize;
+		uint32_t indexAllocationOffset;
+		uint32_t indexAllocationReservedSize;
+		uint32_t vertexAllocationOffset;
+		uint32_t vertexAllocationReservedSize;
+	}*/
 
-		if (i == MDIStructsNeeded - 1)
-		{
-			idxCnt = static_cast<uint32_t>(meshBuffer->getIndexCount()) % maxIdxCntPerPatch;
+	//set vertex data
+	//memcpy()
 
-			if (idxCnt == 0)
-				idxCnt = maxIdxCntPerPatch;
-		}
-		else
+	for (uint64_t i = 0; i < MDIStructsNeeded; i++)
+	{
+		idxBuffBatchEnd = idxBuffBatchBegin + 
+			((i == (MDIStructsNeeded - 1)) ? currMeshBufferIdxCnt % m_maxTriangleCountPerMDIData : m_maxTriangleCountPerMDIData);
+		const uint32_t minIdxVal = *std::min_element(idxBuffBatchBegin, idxBuffBatchEnd);
+
+		MDIStructType MDIData;
+		MDIData.count = idxBuffBatchEnd - idxBuffBatchBegin;
+		MDIData.instanceCount = inputMeshBuffer->getInstanceCount();
+		MDIData.firstIndex = firstIdxForCurrMeshBatch;
+		MDIData.baseVertex = *std::min_element(idxBuffBatchBegin, idxBuffBatchEnd);
+		MDIData.baseInstance = 0u; /*TODO*/
+
+		//set mdi structs
+		memcpy(mdiBuffPtr, &MDIData, sizeof(MDIStructType));
+		
+		//set indices
+		//TODO?: if constexpr for EIT
+		for (IndexType* ptr = idxBuffBatchBegin; ptr != idxBuffBatchEnd; ptr++)
 		{
-			idxCnt = maxIdxCntPerPatch;
+			*indexBuffPtr = *ptr - minIdxVal;
+			indexBuffPtr++;
 		}
 
-		*(mdiBuffPtr + MDIAllocAddr + i) =
-		{
-			idxCnt,
-			static_cast<uint32_t>(meshBuffer->getInstanceCount()),
-			idxAllocAddr + i * maxIdxCntPerPatch,
-			0 /*TODO*/,
-			static_cast<uint32_t>(meshBuffer->getBaseInstance()) /*TODO*/
-		};
-	}
+		firstIdxForCurrMeshBatch += MDIData.count;
+		idxBuffBatchBegin = idxBuffBatchEnd;
 
+		mdiBuffPtr++;
+		indexBuffPtr += MDIData.count;
+	}
 
-	PackedMeshBufferData result{ MDIAllocAddr, MDIStructsNeeded };
-	return result;
+	return {};
 }
 
 template <typename MDIStructType>
-void CCPUMeshPacker<MDIStructType>::commit()
+template <typename Iterator>
+PackedMeshBufferData CCPUMeshPacker<MDIStructType>::commit(const Iterator begin, const Iterator end, ReservedAllocationMeshBuffers ramb)
 {
-	return;
+	//AFTER SAFE SHRINK FIX TODO LIST:
+	//multiple ReservedAllocationMeshbuffers on input
```

Comment:
nope, just one, one `ReservedAllocationMeshbuffers` has the reservation for multiple meshbuffers already